central/delegatedregistryconfig/delegator/delegator.go

```diff
@@ -0,0 +1,148 @@
+		scanWaiterManager: scanWaiterManager,
+	}
+}
+
+type delegatorImpl struct {
+	// deleRegConfigDS for pulling the current delegated registry config.
+	deleRegConfigDS datastore.DataStore
+
+	// connManager for sending scan requests to secured clusters and ensuring
+	// clusters are valid for delegation.
+	connManager connection.Manager
+
+	// scanWaiterManager creates waiters that wait for async scan responses.
+	scanWaiterManager waiter.Manager[*storage.Image]
+}
+
+// GetDelegateClusterID returns the cluster id that should enrich this image (if any) and
+// true if enrichment should be delegated to a secured cluster, false otherwise.
+func (d *delegatorImpl) GetDelegateClusterID(ctx context.Context, image *storage.Image) (string, bool, error) {
+	config, err := d.getConfig(ctx)
+	if err != nil {
+		return "", false, err
+	}
+
+	shouldDelegate, clusterID := d.shouldDelegate(image, config)
+	if !shouldDelegate {
+		return "", false, nil
+	}
+
+	err = d.validateCluster(clusterID)
+	return clusterID, true, err
+}
+
+// DelegateEnrichImage sends an enrichment request to the provided cluster.
+func (d *delegatorImpl) DelegateEnrichImage(ctx context.Context, image *storage.Image, clusterID string, force bool) error {
+	if clusterID == "" {
+		return errors.New("missing cluster id")
+	}
+
+	w, err := d.scanWaiterManager.NewWaiter()
+	if err != nil {
+		return err
+	}
+
+	msg := &central.MsgToSensor{
+		Msg: &central.MsgToSensor_ScanImage{
+			ScanImage: &central.ScanImage{
+				RequestId: w.ID(),
+				ImageName: image.GetName().GetFullName(),
+				Force:     force,
+			},
+		},
+	}
+
+	err = d.connManager.SendMessage(clusterID, msg)
+	if err != nil {
+		w.Close()
+		return err
+	}
+
+	log.Infof("Sent scan request %q to cluster %q for %q", w.ID(), clusterID, image.GetName().GetFullName())
+
+	img, err := w.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("error delegating scan to cluster %q for %q: %w", clusterID, image.GetName().GetFullName(), err)
+	}
+
+	log.Debugf("Scan response received for %q and image %q", w.ID(), img.GetName().GetFullName())
+
+	// Copy the fields from img into image, callers expecting image to be modified in place.
+	*image = *img
+
+	return nil
+}
+
+func (d *delegatorImpl) getConfig(ctx context.Context) (*storage.DelegatedRegistryConfig, error) {
+	config, _, err := d.deleRegConfigDS.GetConfig(ctx)
+	return config, err
+}
+
+func (d *delegatorImpl) shouldDelegate(image *storage.Image, config *storage.DelegatedRegistryConfig) (bool, string) {
+	if config == nil || config.GetEnabledFor() == storage.DelegatedRegistryConfig_NONE {
+		return false, ""
+	}
+
+	var should bool
+	if config.GetEnabledFor() == storage.DelegatedRegistryConfig_ALL {
+		should = true
+	}
+
+	clusterID := config.GetDefaultClusterId()
+	imageFullName := urlfmt.TrimHTTPPrefixes(image.GetName().GetFullName())
+	for _, reg := range config.GetRegistries() {
+		regPath := urlfmt.TrimHTTPPrefixes(reg.GetPath())
+		if strings.HasPrefix(imageFullName, regPath) {
```

Comment:
This Delegator logic has changed based on the code review requests above (haven't pushed yet). Now the Delegator receives a `storage.ImageName` instead of a mostly empty `storage.Image`.  So `image.GetNames()` is not avail anymore.

The `Names` field would have always been empty however because this was being called from ad-hoc flows that would only have the name supplied in the request (`roxctl image scan --image=blah` for example). 

There are multiple points in the workflow after delegation that match the image to registries (such as [here](https://github.com/stackrox/stackrox/blob/f1abdaf371b14126c746530d96a2d634eec4401e/sensor/common/scan/scan.go#L155) and [here](https://github.com/stackrox/stackrox/blob/f1abdaf371b14126c746530d96a2d634eec4401e/central/image/service/service_impl.go#L382)) that pull/validate sigs. We may need a revisit these in another PR to include the matching logic you shared.

As a sidebar, I see image.Names populated from the scan requests from Sensor, but not seeing the field populated from the Central workflows or on cache retrievals in the image enricher. Unless I'm missing something each Central workflow may be erasing the Names field when it updates the DB.

Reply:
>The Names field would have always been empty however because this was being called from ad-hoc flows that would only have the name supplied in the request (roxctl image scan --image=blah for example).

It won't be empty but essentially the value within the `Name` field, see [types.ToImage](https://github.com/stackrox/stackrox/blob/f1abdaf371b14126c746530d96a2d634eec4401e/pkg/images/types/type.go#L12) (AFAICT this is what the detection service also calls).

>There are multiple points in the workflow after delegation that match the image to registries (such as here and here) that pull/validate sigs.

Yeah, true that.

>As a sidebar, I see image.Names populated from the scan requests from Sensor, but not seeing the field populated from the Central workflows or on cache retrievals in the image enricher. Unless I'm missing something each Central workflow shown in the table at the top of this PR may be erasing the Names field when saving to DB.

Yeah the initial focus was on the sensor <-> scanner workflow, roxctl didn't matter that much in that case (but probably should as well). Looks like the fetch from database doesn't populate that field tho, I'll fix that.